plone/app/vocabularies/__init__.py

```diff
@@ -32,3 +34,28 @@ def __getitem__(self, start, stop=None):
 
     def __len__(self):
         return len(self._terms)
+
+
+@implementer(IPermissiveVocabulary)
+class PermissiveVocabulary(SimpleVocabulary):
+    """
+    Permissive vocabulary for cases of integer-keyed choices or cases
+    where vocabulary may mutate later in a transaction to include a
+    newly inserted value.
+    """
+
+    def __contains__(self, value):
+        return True
+
+    def getTermByToken(self, token):
+        """
+        this works around z3c.form.widget.SequenceWidget.extract()
+        pseudo-validation (which is broken for a permissive vocabulary).
+        """
+        try:
+            v = super(PermissiveVocabulary, self).getTermByToken(token)
+        except LookupError:
+            # fallback using dummy term, assumes token==value
+            return SimpleTerm(token)
```

Comment:
FWIW, there are some vocabularies that use more exotic term value types in this package (tuple, catalog brain), and those are outside the scope or effect of this change, so this should be safe (and in those cases it is not use of token-as-value that is problem, IIRC, but value-as-token).

I believe that in all cases where PermissiveVocabulary subclass is used, it is safe to use token as a value -- especially in this case, as getTermByToken() is only used here to make a z3c.form widget happy for values not (yet) contained.